packages/peregrine/lib/talons/CheckoutPage/PaymentInformation/usePaymentInformation.js

```diff
@@ -57,71 +55,107 @@ export const usePaymentInformation = props => {
 
-    const onPaymentDetailsQueryCompleted = useCallback(
-        paymentDetailsData => {
-            /**
-             * If there is a selected payment method saved on cart,
-             * and paymentNonceDetails in cache, it means the payment
-             * information step is completed. Call onSave to move to next step.
-             *
-             * Do not call if paymentNonceDetails is not in cache, because we
-             * need it to render the summary. There is no way to save it on
-             * the remote server, hence we save it in local cache.
-             */
-            const selectedPaymentMethodOnCart =
-                paymentDetailsData &&
-                paymentDetailsData.cart.selectedPaymentMethod
-                    ? paymentDetailsData.cart.selectedPaymentMethod.code
-                    : null;
-            const paymentNonceDetails = paymentDetailsData
-                ? paymentDetailsData.cart.paymentNonce
-                : null;
-
-            const hasData = !!(
-                selectedPaymentMethodOnCart && paymentNonceDetails
-            );
-
-            if (hasData) {
-                if (onSave) {
-                    onSave();
-                }
-            }
-
-            setHasData(hasData);
-            setSkipQueryCall(true);
-        },
-        [onSave]
-    );
-
     /**
      * Queries
      */
+    const {
+        data: paymentInformationData,
+        loading: paymentInformationLoading
+    } = useQuery(getPaymentInformation, {
+        variables: { cartId }
+    });
+    const [
+        setPaymentMethod,
+        { loading: setPaymentMethodLoading }
+    ] = useMutation(setPaymentMethodMutation);
+    /**
+     * Effects
+     */
 
-    useQuery(getPaymentDetailsQuery, {
-        variables: { cartId },
-        skip: skipQueryCall,
-        onCompleted: onPaymentDetailsQueryCompleted
+    const availablePaymentMethods = paymentInformationData
+        ? paymentInformationData.cart.available_payment_methods
+        : [];
+
+    const selectedPaymentMethod =
+        (paymentInformationData &&
+            paymentInformationData.cart.selected_payment_method.code) ||
+        null;
+
+    // Whenever available methods change we should reset to the editing view
+    // so that a user can see the newly available methods. This could occur
+    // if a user causes their cart total to become $0. Additionally, the
+    // current pattern requires that the method components themselves
+    // indicate their "done" state so we must leave it to them to revert
+    // this effect downstream.
+    useEffect(() => {
+        setDoneEditing(false);
+    }, [availablePaymentMethods]);
+
+    // Along the lines of the above effect, since we don't have a "free method"
+    // component we check if free is available and set it as the selected method
+    // directly. What results is a potential flash of selectable payment methods
+    // and then the "free" summary.
+    useEffect(() => {
```

Comment:
As @revanth0212 showed in his commit it is possible to have free payment as a method and separate this effect out into a component. See here: https://github.com/magento/pwa-studio/compare/rugh/pwa-563-free-payment-method...rugh/pwa-563-free-payment-method-as-method?expand=1

I'm not really sure which method I prefer but we can discuss the pros/cons here.

Reply:
Thanks for the diff @sirugh. I would prefer it to be a separate payment method to make the payment workflow generic for all current and future methods.

Pros:
1. Generic workflow for all payment methods. All payment methods get the same props and they deal with how to submit the info and let the parent know when they are done.
2. Leaner Payment Info component and talon. Easy to maintain and scale.

Cons:
1. Rework

Reply:
To add to @revanth0212's pros and cons, the current implementation allows for a generic workflow for all "normal" payment methods right now. A new component will receive the same props as braintree and should "submit" when `shouldSubmit` or call `setDoneEditing(true)` when they don't need to render input forms. Free is a special case and needs to be applied as soon as it becomes available so it actually does not fit into the UX in the same way.

In the comparison I linked above the free payment method works as desired but there is additional work to be done within the credit card component to have the solution work for all the use cases I listed in the verification steps.